### ISSUE TITLE
undefined symbol: napi_module_register

### ISSUE THREAD

**Issue by edisionnano** [Mysigyeong/analyze-discord#1]
I compiled harness.cc using `g++ harness.cc -o harness` after changing the path in line 44 from /root/ to `/snap/discord/132/.config/discord/0.0.17/modules/discord_voice/` as that's where the file is located on my Ubuntu machine.
<br>When I run the file using `./harness file` I get `./harness: symbol lookup error: /home/spiros/snap/discord/132/.config/discord/0.0.17/modules/discord_voice/discord_voice.node: undefined symbol: napi_module_register`

**Comment by Mysigyeong** [comment]
You have to preload the libraries by using LD_PRELOAD. I preloaded libnode.so, libavcodec.so, libXcomposite.so, libXrandr.so, and libXdamage.so. The offset of function can be different if the library is updated. Please check before running the harness.

**Comment by edisionnano** [comment]
> You have to preload the libraries by using LD_PRELOAD. I preloaded libnode.so, libavcodec.so, libXcomposite.so, libXrandr.so, and libXdamage.so. The offset of function can be different if the library is updated. Please check before running the harness.

Thanks, I was able to get further after preloading libnode.so. I also preloaded libavcodec.so.58. I am now getting a segfault.
<br>`decoder size = -4

Program received signal SIGSEGV, Segmentation fault.
0x00007fffec396aa0 in opus_decoder_init () from /home/spiros/snap/discord/132/.config/discord/0.0.17/modules/discord_voice/discord_voice.node`
<br>Perhaps the offset changed?

**Comment by Mysigyeong** [comment]
Yeah, maybe there was an update after I made the harness... So, offset might be changed. I used the offset to call the functions because the functions cannot be loaded directly with dlsym function.

**Comment by edisionnano** [comment]
> Yeah, maybe there was an update after I made the harness... So, offset might be changed. I used the offset to call the functions because the functions cannot be loaded directly with dlsym function.

How do you find the offset?

**Comment by edisionnano** [comment]
> I did reversing to find the offset. I used ghidra to analyze the library.

I see, thanks. I was able to find the new offsets of the symbols using the nm command. For example I did `nm discord_voice.node | grep opus_decode` to get `00000000002ef020 t opus_decode`. Now I have to see if 0x2ef020 is the correct offset for the pointer

**Comment by edisionnano** [comment]
Ok so I've replaced the new offsets
```
const int ERROR_KIND_OFFSET = 0xb29d50;
const int OPUS_DECODE_OFFSET = 0x2ef020;
const int OPUS_DECODER_CREATE_OFFSET = 0x2ecac0;
const int OPUS_DECODER_CTL_OFFSET = 0x2ef1f0;
const int OPUS_DECODER_GET_SIZE_OFFSET = 0x2ec8f0;
```
And got an opus file from https://filesamples.com/samples/audio/opus/sample4.opus
<br>I run the harness using `LD_PRELOAD="libnode.so libavcodec.so.58" ./harness sample4.opus` and I get
```
decoder size = 26996
data len = 3420150
decoded = -4
```
I'm assuming -4 is an error code
